### PR TITLE
ezpz/moe: resync with upstream + adapt to PR #3308 (use_grouped_mm removal)

### PR DESCRIPTION
**Pull request by saforem2**
## Summary

Pulls in 12 commits from `upstream/main` and replays the only ezpz-affecting change onto `experiments/ezpz/moe/`. Two-commit history; no expert-side perf work or HSDP fixes here — those stay with the open PRs (#9, #10, #11) which will need to rebase onto this.

## Commits

1. **`62b395b5` — Merge `upstream/main` into `ezpz-moe-resync`.** Clean merge (0 conflicts pre-checked). Notable upstream changes:
   - `b301dfa0` — **[MoE] Remove expert for-loop fallback (#3308)** — deletes `_run_experts_for_loop` and the `use_grouped_mm` config field from `models/common/moe.py`. Inlines `torch._grouped_mm` as the only expert path. **This breaks `experiments/ezpz/moe/model.py`** which used `use_grouped_mm = False` as the XPU fallback (XPU has no `_grouped_mm` kernel at all).
   - `5ca23a5d` — [GraphTrainer] Add Context Parallel support
   - `d57df092` — Make ChunkedCELoss support `torch.autograd.grad`
   - `1a0fe3e3` — [graph_trainer] Refactor passes.py into focused modules
   - …plus 8 more (graph_trainer / RL / lint / docs)

2. **`12aba852` — `feat(ezpz/moe): introduce EzpzGroupedExperts to handle upstream #3308`**

   - **New `experiments/ezpz/moe/experts.py`**: `EzpzGroupedExperts(GroupedExperts)` subclass with `compute_backend: Literal[\"for_loop\", \"grouped_mm\"]`. Default `\"grouped_mm\"` defers to upstream via `super()._experts_forward(...)`. `\"for_loop\"` re-vendors the per-expert matmul loop that #3308 deleted, restoring the XPU / pre-SM90 path.

   - **`experiments/ezpz/moe/__init__.py`**: new `make_ezpz_experts_config(...)` wrapper that calls upstream's `make_experts_config(...)` and re-wraps as `EzpzGroupedExperts.Config`. `_build_moe_layers` now threads a `compute_backend` kwarg (default `\"grouped_mm\"`) through to it. All existing flavors (`debugmodel`, `500M`, `2B`, `10B_2B`, `10B_2B_sdpa`, …) build unchanged.

   - **`experiments/ezpz/moe/model.py`**: `update_from_config` previously mutated `experts.use_grouped_mm = False` on pre-SM90 devices. That field no longer exists. Replaced with `experts_cfg.compute_backend = \"for_loop\"`, guarded by a `getattr(..., \"compute_backend\", \"grouped_mm\")` so the block is robust to future config-shape changes.

   - **`experiments/ezpz/docs/upstream-sync.md`**: added the 33rd-sync entry per the upstream-sync protocol, documenting #3308 and the replay.

## Default behavior is unchanged

For SM90+ CUDA, `compute_backend` defaults to `\"grouped_mm\"`, which calls `super()._experts_forward(...)` and hits the upstream path. The new `for_loop` branch only fires on devices without `_grouped_mm` (currently XPU and pre-SM90 CUDA).

End-to-end config build verified locally for `debugmodel`, `500M`, `10B_2B_sdpa` — all instantiate `EzpzGroupedExperts` with `compute_backend=\"grouped_mm\"`.

## Notes for downstream PRs

PRs #9, #10, #11 all assume the pre-#3308 `GroupedExperts` shape:

- **#9 (Sam Wheeler — HSDP fix)**: doesn't touch `moe.py`, but lives on top of pre-merge `parallelize.py`. Should rebase cleanly.
- **#10 (Sam Wheeler — `batched_mm_padded` backend)**: extends `GroupedExperts.Config` with a `compute_backend` field that's basically the same shape as the one this PR adds via subclass. Cleanest landing path is to add `\"batched_mm_padded\"` as a third literal in `EzpzGroupedExperts.ExpertComputeBackend` and a third branch in `EzpzGroupedExperts._experts_forward`.
- **#11 (Nathan Nichols — MoE optimizations)**: expert-side optimizations (cached `w13` / `w2_t` + equal-counts no-grad bmm fast path) layer onto the `for_loop` method here. Token-dispatcher rewrite is unaffected by this PR.

Will leave pointer comments on #9 / #10 / #11.

## Out of scope

This PR deliberately does NOT bundle the changes from open PRs #9, #10, #11. Those are owned by Sam and Nathan respectively and should land independently after rebasing onto this. Supersedes #12 (the prior 4-commit integration attempt) which will be closed.

## Validation

- [x] Clean merge of `upstream/main` (0 conflicts, pre-checked with `git merge-tree`)
- [x] All existing MoE flavors build with the new `EzpzGroupedExperts.Config` wrapper
- [x] No remaining references to `use_grouped_mm` in `experiments/ezpz/` (verified via grep)
- [ ] MoE smoke run on Aurora / Sunspot (XPU `for_loop` path) — to be filed as a follow-up smoke-test report under `experiments/ezpz/docs/experiments/moe/`

### CODE REVIEW

File: torchtitan/experiments/graph_trainer/chunked_loss.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        accumulated_h_grad: torch.Tensor,
+        total_loss: torch.Tensor,
+        lm_head: nn.Module,
+        fsdp_enabled: bool,
+        *lm_params: torch.Tensor,
+    ) -> torch.Tensor:
+        # The chunk loop above already populated each lm_head param's
+        # ``.grad`` with the correctly sharded value via the FSDP last-chunk
+        # post-accumulate-grad hook (reduce-scatter). Capture those grads
+        # into saved_tensors so backward ca route them as autograd outputs
```
[the remark this answers is not in the document]
Comment (saforem2) [reply torchtitan/experiments/graph_trainer/chunked_loss.py:76]: Same — `experiments/graph_trainer/chunked_loss.py` is upstream code, brand new in `1a0fe3e3` (pytorch/torchtitan#3319) and pulled in via the upstream merge. Not modifying it here; would need a pytorch/torchtitan PR.

File: torchtitan/components/loss.py [hunk 2]
```diff
@@ -387,4 +402,8 @@ def backward(  # pyrefly: ignore[bad-override]
         # decoder graph — equivalent to hidden_states.backward(accumulated_grad)
         # but expressed as a return value so autograd handles the traversal
         # in a single pass (no "backward through graph twice" error).
+        # Note: this is not safe if downstream accidentally runs tensor ops after
+        # the loss returns, which would produce a non-trivial grad_output that we need
+        # to properly handle. The complicated part is that grad_output might not be
+        # on the same device mesh as accumlated_grad.
```
[the remark this answers is not in the document]
Comment (saforem2) [reply torchtitan/components/loss.py:408]: Good catch, but this lives in `torchtitan/components/loss.py` (upstream core) — it landed via the `git merge upstream/main` in `62b395b5`. Per the ezpz golden rule we don't modify code outside `experiments/ezpz/` in this fork; the fix belongs in a pytorch/torchtitan PR. Leaving as-is here.
